src/Microsoft.AspNet.OData.Shared/Builder/ODataConventionModelBuilder.cs

```diff
@@ -1080,21 +1080,48 @@ private void ApplyPropertyConvention(IEdmPropertyConvention propertyConvention,
             }
         }
 
-        private static Dictionary<Type, List<Type>> BuildDerivedTypesMapping(IWebApiAssembliesResolver assemblyResolver)
+        private static Dictionary<Type, Type[]> BuildDerivedTypesMapping(IWebApiAssembliesResolver assemblyResolver)
         {
-            IEnumerable<Type> allTypes = TypeHelper.GetLoadedTypes(assemblyResolver).Where(t => TypeHelper.IsVisible(t) && TypeHelper.IsClass(t) && t != typeof(object));
-            Dictionary<Type, List<Type>> allTypeMapping = allTypes.Distinct().ToDictionary(k => k, k => new List<Type>());
-
-            foreach (Type type in allTypes)
+            Predicate<Type> verifyType = t => TypeHelper.IsVisible(t) && TypeHelper.IsClass(t) && t != typeof(object);
+
+            // The dictionary is allocated to contain all visible reference types which is not system.Object, each with an List<Type> allocated.
+            // The list is used to keep derived types.
+            // In a common scenario, there are about 10% types with derived types.
+            // So, we don't need to keep a large number of empty list in the dictionary.
+            Dictionary<Type, List<Type>> temp = new Dictionary<Type, List<Type>>();
+            foreach (Type type in TypeHelper.GetLoadedTypes(assemblyResolver))
             {
-                List<Type> derivedTypes;
-                if (TypeHelper.GetBaseType(type) != null && allTypeMapping.TryGetValue(TypeHelper.GetBaseType(type), out derivedTypes))
+                if (type != null && verifyType(type))
                 {
-                    derivedTypes.Add(type);
+                    Type baseType = TypeHelper.GetBaseType(type);
+                    if (baseType != null)
+                    {
+                        List<Type> list;
+                        if (!temp.TryGetValue(baseType, out list))
```

Comment:
@xuzhg There might be a bug here...
Supposing `GetLoadedTypes` method returned the types in the following order:
1. `A : Object` - _Object is the base type of A_
2. `C : B` - _B is the based type of C_
3. `B : A` - _A is the base type of B_

When "processing" type `C`, the `TryGetValue` call here will evaluate to false and we'll end up not establishing the relationship between `B` and `C`.

Maybe this can be remedied by also checking for any types derived from `type` that might already be in the `temp` dictionary and adding them to `temp[type]` list

Reply:
Unless `GetLoadedTypes()` always returns base types ahead of derived types...

Reply:
Thanks @xuzhg for the explanation. Makes sense. That we're using `baseType` as the key and adding `type` to the `list` (value) is the statement that didn't register in my mind properly